Model/Indexer/Product/Indexer.php

```diff
@@ -58,22 +96,49 @@ public function __construct(
      */
     public function executeFull()
     {
-        if ($this->dataHelper->isFullReindexEnabled()) {
-            $pageNumber = 0;
+        if ($this->dataHelper->isFullReindexEnabled()
+            && !empty($this->nostoHelperAccount->getStoresWithNosto())
+        ) {
+            // Truncate queue table before first execution if they have leftover products
+            if ($this->nostoQueueRepository->isQueuePopulated()) {
+                $this->nostoQueueRepository->truncate();
+            }
+            $productCollection = $this->getProductCollection();
+            $productCollection->setPageSize(self::BATCH_SIZE);
+            $lastPage = $productCollection->getLastPageNumber();
+            $pageNumber = 1;
             do {
-                $pageNumber++;
-                $searchCriteria = $this->searchCriteriaBuilder
-                    ->addFilter('status', Status::STATUS_ENABLED, 'eq')
-                    ->setPageSize(self::BATCH_SIZE)
-                    ->setCurrentPage($pageNumber)
-                    ->create();
-                $products = $this->productRepository->getList($searchCriteria);
-                if ($products instanceof SearchResultsInterface) {
-                    $this->productService->update($products->getItems());
+                $productCollection->setCurPage($pageNumber);
+                $productCollection->addAttributeToSelect('id')
+                    ->addAttributeToFilter(
+                        'status',
+                        ['eq'=> Status::STATUS_ENABLED]
+                    )->addAttributeToFilter(
+                        'visibility',
+                        ['neq'=> Visibility::VISIBILITY_NOT_VISIBLE]
+                    );
+                $products = [];
+                foreach ($productCollection->getItems() as $product) {
+                    $products[$product->getId()] = $product->getTypeId();
                 }
-            } while (($pageNumber * self::BATCH_SIZE) <= $products->getTotalCount());
+                $productCollection->clear();
+                $this->logger->logWithMemoryConsumption(
+                    sprintf('Indexing from executeFull, remaining pages: %d', $lastPage - $pageNumber)
+                );
+                try {
+                    $this->productService->update($products);
+                } catch (MemoryOutOfBoundsException $e) {
+                    throw new LocalizedException(new Phrase($e->getMessage()));
+                }
+                $this->productService->processed = [];
+                $products = null;
+                $pageNumber++;
+            } while ($pageNumber <= $lastPage);
+            $productCollection = null;
         } else {
-            $this->logger->info('Skip full reindex since full reindex is disabled.');
+            $this->logger->info(
+                'Skip full reindex since full reindex is disabled or Nosto account is not connected.'
```

Comment:
Tiny change to the log message 'Skip full reindex since full reindex is disabled or Nosto account is not connected into any store view.'